src/NovelAudioService.cpp

```diff
@@ -0,0 +1,367 @@
+  {
+    std::cout << "ERROR: Failed to allocate channels." << std::endl;
+  }
+  else
+  {
+    std::cout << "INFO: Mixer engaged!" << std::endl;
+    isInitialized = true;
+  }
+  return isInitialized;
+}
+
+void NovelAudioService::loadSound(std::string input) {
+  auto exists = _sounds.find(input);
+  if (exists != _sounds.end()) return;
+
+  Mix_Chunk* newSound = Mix_LoadWAV(input.c_str());
+  if (newSound != nullptr)
+  {
+    _sounds[input] = newSound;
+  }
+  else
+  {
+    std::cerr << Mix_GetError() << std::endl;
+  }
+
+
+}
+
+void NovelAudioService::unloadSound(std::string input) {
+  auto existingSound = _sounds.find(input);
+  if (existingSound == _sounds.end()) return;
+
+  Mix_FreeChunk(_sounds[input]);
+  _sounds.erase(existingSound);
+}
+
+void NovelAudioService::playSound(std::string soundName) {
+  auto existingSound = _sounds.find(soundName);
+  if (existingSound == _sounds.end())
+  {
+    loadSound(soundName);
+    existingSound = _sounds.find(soundName);
+    if (existingSound == _sounds.end())
+    {
+      return;
+    }
+  }
+
+  auto onChannel = _channelMap.find(soundName);
+  if (onChannel == _channelMap.end())
+  {
+    _channelMap[soundName] = _nextChannel;
+    Mix_PlayChannel(_nextChannel, _sounds[soundName], 0);
+    int nextChannelTest = _nextChannel + 1;
+    _nextChannel = (nextChannelTest >= 2 || nextChannelTest < 0) ? 0 : nextChannelTest;
+  }
+  else
+  {
+    if (Mix_Playing(_channelMap[soundName]) == 1)
+    {
+      std::cout << "Already playing on channel " << _channelMap[soundName] << std::endl;
+      return;
+    }
+
+    Mix_PlayChannel(_channelMap[soundName], _sounds[soundName], 0);
+  }
+
+}
+
+void NovelAudioService::playSound(std::string soundName, int loops) {
+  auto existingSound = _sounds.find(soundName);
+  if (existingSound == _sounds.end())
+  {
+    loadSound(soundName);
+    existingSound = _sounds.find(soundName);
+    if (existingSound == _sounds.end())
+    {
+      return;
+    }
+  }
+
+  auto onChannel = _channelMap.find(soundName);
+  if (onChannel == _channelMap.end())
+  {
+    _channelMap[soundName] = _nextChannel;
+    Mix_PlayChannel(_nextChannel, _sounds[soundName], loops-1);
+    int nextChannelTest = _nextChannel + 1;
+    _nextChannel = (nextChannelTest >= 2 || nextChannelTest < 0) ? 0 : nextChannelTest;
+  }
+  else
+  {
+    if (Mix_Playing(_channelMap[soundName]) == 1)
+    {
+      std::cout << "Already playing on channel " << _channelMap[soundName] << std::endl;
+      return;
+    }
+    Mix_PlayChannel(_channelMap[soundName], _sounds[soundName], loops-1);
+  }
+
+}
+
+void NovelAudioService::stopSound(std::string soundName) {
+  Mix_HaltChannel(_channelMap[soundName]);
+}
+
+int NovelAudioService::getSoundChannel(std::string soundName) {
+  auto existingSound = _channelMap.find(soundName);
+  if (existingSound == _channelMap.end())
+  {
+    std::cerr << "ERROR: Cannot get channel for sound " << soundName << std::endl;
+    return -9;
```

Comment:
Why `-9`?

Reply:
To be honest.... -3 is an internally used number (exclusive to NovelRT), -2 is the Mixer's postmix channel, -1 is Mixer's any/all channel definition, and anything 0 or positive can be a possible channel. Hence (-9) - it'll never be accepted by Mixer, and it's easily identifiable by us. :D
(in retrospect, I'll probably change this since it's random. Just realized it was defined as -9 in the newest commit too >.<)

FYI @RubyNova

Reply:
Ahahahahaha oh my lord that's amazing.

Maybe make a `const` for it if you absolutely must?

Reply:
Newest commit already did, but I'm going to remove it and change it tbh. Re-reading the code, if I find a `-1` with this function, it should return the error as a channel hasn't actually been assigned yet.

Reply:
Removed in latest commit 👍 (after cleaning up history)